gobblin-data-management/src/test/java/org/apache/gobblin/data/management/copy/iceberg/IcebergDatasetTest.java

```diff
@@ -216,210 +228,295 @@ public void testPathErrorConsolidator() {
-    Arrays.stream(snapshotDefs).flatMap(snapshotDef ->
-        snapshotDef.asSnapshotInfo().getAllPaths().stream()
-    ).forEach(p ->
-        paths.add(new Path(p))
-    );
+    Arrays.stream(snapshotDefs).flatMap(snapshotDef -> snapshotDef.asSnapshotInfo().getAllPaths().stream())
+        .forEach(p -> paths.add(new Path(p)));
     return paths;
   }
 
   private CopyConfiguration createEmptyCopyConfiguration(FileSystem fs) {
-    return CopyConfiguration.builder(fs, copyConfigProperties)
-        .copyContext(new CopyContext())
-        .build();
+    return CopyConfiguration.builder(fs, copyConfigProperties).copyContext(new CopyContext()).build();
   }
 
-  private static void verifyCopyEntities(Collection<CopyEntity> copyEntities, List<String> expected) {
-    List<String> actual = new ArrayList<>();
+  private static void verifyCopyEntities(Collection<CopyEntity> copyEntities, Set<String> expected) {
+    Set<String> actual = new HashSet<>();
+    CopyEntityDeserializer copyEntityDeserializer = new CopyEntityDeserializer();
     for (CopyEntity copyEntity : copyEntities) {
       String json = copyEntity.toString();
-      String filepath = new Gson().fromJson(json, JsonObject.class)
-          .getAsJsonObject("object-data").getAsJsonObject("origin")
-          .getAsJsonObject("object-data").getAsJsonObject("path")
-          .getAsJsonObject("object-data").getAsJsonObject("uri")
-          .getAsJsonPrimitive("object-data").getAsString();
+      String filepath = copyEntityDeserializer.getFilePathAsStringFromJson(json);
       actual.add(filepath);
     }
-    Assert.assertEquals(actual.size(), expected.size(),
-        "Set" + actual.toString() + " vs Set" + expected.toString());
+    Assert.assertEquals(actual.size(), expected.size(), "Set" + actual.toString() + " vs Set" + expected.toString());
     Assert.assertEqualsNoOrder(actual.toArray(), expected.toArray());
   }
 
+  private void verifyFsOwnershipAndPermissionPreservation(Collection<CopyEntity> copyEntities,
+      Map<String, FileStatus> expectedMap) {
+    CopyEntityDeserializer copyEntityDeserializer = new CopyEntityDeserializer();
+
+    for (CopyEntity copyEntity : copyEntities) {
+      String copyEntityJson = copyEntity.toString();
+
+      JsonArray ancestorOwnerAndPermissions = copyEntityDeserializer.getAncestorOwnerAndPermissions(copyEntityJson);
+      String filePath = copyEntityDeserializer.getFilePathAsStringFromJson(copyEntityJson);
+      JsonObject destinationOwnerAndPermissions =
+          copyEntityDeserializer.getDestinationOwnerAndPermissions(copyEntityJson);
+
+      FileStatus expected = expectedMap.get(filePath);
+      Map<String, String> actual =
+          copyEntityDeserializer.getDestinationOwnerAndPermissionsFsPermissionsMap(destinationOwnerAndPermissions);
+
+      Assert.assertEquals(ancestorOwnerAndPermissions.size(), new Path(filePath).getParent().depth() - 1);
```

Comment:
shan't we verify more than the length?

BTW, where are the user and group names verified?

Reply:
seems we still don't verify the specific values along the ancestor chain...